### ISSUE TITLE
_set_seed() mutates Python/numpy/torch global random state — user reproducibility broken silently, no isolation between concurrent models

### ISSUE THREAD

**Issue by Afnan-abdi** [dataxid/dataxid-python#187]
## Description
When a user passes seed=42 to ModelConfig, the SDK calls _set_seed(42) in Model.create() . The implementation in dataxid/training/_seed.py is:
```python
pythondef _set_seed(seed: int) -> None:
    random.seed(seed)            # mutates Python's global RNG
    np.random.seed(seed)         # mutates numpy's global legacy RNG
    torch.manual_seed(seed)      # mutates PyTorch's global CPU RNG
    torch.cuda.manual_seed_all(seed)  # mutates all CUDA RNG states
```

Every one of these calls mutates global process-level state. This creates three distinct problems:
1. Silent contamination of user code. A user who runs model = dataxid.Model.create(data, config=ModelConfig(seed=42)) will find that their subsequent calls to random.random(), np.random.choice(...), or torch.randn(...) produce different values than before the DataXID call. Users expect an SDK to manage its own reproducibility without reaching into their process state. This is a silent, hard-to-debug interaction between DataXID and any other seeded library the user runs in the same notebook.
2. No isolation between concurrent models. Running two Model.create(seed=X) calls in parallel threads (a reasonable thing to want to do on multi-core machines) produces a race on the global state. Each call overwrites the other's seed mid-training. The resulting random draws during encoding are non-deterministic despite both models being seeded.
3. Other internal call sites use global state directly. Separate from _set_seed, the SDK also reaches into global random state from:

dataxid/pipeline/_encode.py:56 — np.random.choice(...) for NaN imputation
dataxid/pipeline/_decode.py:78, 105, 137 — np.random.choice and np.random.uniform for rare-category handling and bin sampling
dataxid/encoder/_wrapper.py:314, 384 — random.shuffle(all_indices) for training batch order

None of these take an explicit generator; all of them inherit whatever the global state happens to be. This is classic "action at a distance" — the randomness of encoding and shuffling silently depends on whether and when _set_seed was called earlier in the process.

## Steps to Reproduce
Problem 1 — user code contamination:
```python
import random, numpy as np, dataxid
import pandas as pd

# Before: the user establishes their own reproducible pipeline
random.seed(7)
my_list = [random.random() for _ in range(3)]
# [0.3229..., 0.0168..., 0.8356...]

# ... time passes ...
# ... user calls dataxid ...
dataxid.api_key = "dx_test_dummy"
try:
    dataxid.synthesize(
        data=pd.DataFrame({"x": [1,2,3]}),
        n_samples=5,
        config=dataxid.ModelConfig(seed=42),
    )
except Exception:
    pass  # we only need to observe the side effect

# User's global random state has been silently reseeded to 42
new_list = [random.random() for _ in range(3)]
# Different values. The user's "reproducible" pipeline is now reproducible
# with DataXID's seed, not theirs.
```
Problem 2 — concurrency race:
```python
import concurrent.futures, dataxid, pandas as pd
df = pd.DataFrame({"x": list(range(100))})

def run(seed):
    dataxid.Model.create(df, config=dataxid.ModelConfig(seed=seed))

# Two threads reseed the global state interleavedly;
# neither model is actually trained with its requested seed.
with concurrent.futures.ThreadPoolExecutor(max_workers=2) as pool:
    pool.map(run, [42, 1337])
```
Problem 3 — state leaks in encoding:
```python
# dataxid/pipeline/_encode.py line 56:
values[nan_mask] = np.random.choice(vc.index, size=nan_mask.sum(), p=vc.values)
# This is the imputer; it uses global np.random regardless of ModelConfig.seed.
```

## Expected Behavior
 library should never mutate process-global random state. The fix is to (a) use np.random.default_rng(seed) to construct an isolated Generator per model, and (b) pass that generator explicitly through to every call site that needs randomness. This is the NumPy 1.17+ recommended pattern for library authors. For Python's random, the equivalent is random.Random(seed) (an instance, not the module).

## Actual Behavior
All three problems above manifest. The SDK's reproducibility flag reaches into process state, and its internal randomness pulls from that same process state.

## Configuration

```python
config = {
    #Default. Triggered any time ModelConfig(seed=X) is used, or any time encoding/decoding runs (which is always).
}
```

## Environment

- **dataxid version**: `0.3.0`
- **Python version**: `3.12.13`
- **OS**:Linux 6.6.113+ (Google Colab)
## proposed fix
Replace _set_seed with a factory that returns an isolated RNG, then thread it through:
```python
# dataxid/training/_seed.py
from __future__ import annotations
import numpy as np
import torch

class RngBundle:
    """Isolated random-number generators for a single Model instance."""
    def __init__(self, seed: int | None):
        self.np = np.random.default_rng(seed)
        self.torch = torch.Generator()
        if seed is not None:
            self.torch.manual_seed(seed)

    def shuffle(self, seq: list) -> None:
        """Numpy-backed in-place shuffle that does not touch global state."""
        self.np.shuffle(seq)


def make_rng(seed: int | None) -> RngBundle:
    return RngBundle(seed)
```
Call sites (_encode.py:56, _decode.py:78/105/137, _wrapper.py:314/384) should accept an rng: RngBundle argument and use rng.np.choice(...), rng.shuffle(...), etc., instead of the module globals. The Model instance holds the RngBundle, and it is passed down through the encoding and decoding pipelines.
This is a larger refactor than the other fixes, but it closes the entire class of seed-related bugs  including the silently-ignored seed parameter reported previously by other students  and makes the SDK safe for concurrent use.

**Comment by BurakUlver** [comment]
Nice catch, reproduced on 0.3.1. Passing seed into ModelConfig reseeds process-global random / NumPy / PyTorch, so later user RNG draws can change. We'll want isolated RNGs here; larger change, so tracking for now.
